# Project: Pizza Runner

## Problem Statement
- Pizza Runner is in need of your assistance to help the restaurant grow. 
- The restaurant wants to use the data it has captured to get information about customer ordering habits, how much money the restaurant has made so far, and how efficient their delivery people are.

### Connect to SQL

In [63]:
%%capture
%load_ext sql
%sql sqlite://

## Data Tables

### Customer Orders Table

In [64]:
%%sql
CREATE TABLE customer_orders (
  "order_id" INTEGER,
  "customer_id" INTEGER,
  "pizza_id" INTEGER,
  "exclusions" VARCHAR(4),
  "extras" VARCHAR(4),
  "order_time" TIMESTAMP
);

INSERT INTO customer_orders
  ("order_id", "customer_id", "pizza_id", "exclusions", "extras", "order_time")
VALUES
  ('1', '101', '1', '', '', '2020-01-01 18:05:02'),
  ('2', '101', '1', '', '', '2020-01-01 19:00:52'),
  ('3', '102', '1', '', '', '2020-01-02 23:51:23'),
  ('3', '102', '2', '', NULL, '2020-01-02 23:51:23'),
  ('4', '103', '1', '4', '', '2020-01-04 13:23:46'),
  ('4', '103', '1', '4', '', '2020-01-04 13:23:46'),
  ('4', '103', '2', '4', '', '2020-01-04 13:23:46'),
  ('5', '104', '1', 'null', '1', '2020-01-08 21:00:29'),
  ('6', '101', '2', 'null', 'null', '2020-01-08 21:03:13'),
  ('7', '105', '2', 'null', '1', '2020-01-08 21:20:29'),
  ('8', '102', '1', 'null', 'null', '2020-01-09 23:54:33'),
  ('9', '103', '1', '4', '1, 5', '2020-01-10 11:22:59'),
  ('10', '104', '1', 'null', 'null', '2020-01-11 18:34:49'),
  ('10', '104', '1', '2, 6', '1, 4', '2020-01-11 18:34:49');

 * sqlite://
(sqlite3.OperationalError) table customer_orders already exists
[SQL: CREATE TABLE customer_orders (
  "order_id" INTEGER,
  "customer_id" INTEGER,
  "pizza_id" INTEGER,
  "exclusions" VARCHAR(4),
  "extras" VARCHAR(4),
  "order_time" TIMESTAMP
);]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


In [65]:
%%sql
SELECT *
FROM customer_orders

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02'),
 (2, 101, 1, '', '', '2020-01-01 19:00:52'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23'),
 (3, 102, 2, '', '', '2020-01-02 23:51:23'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46'),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29'),
 (6, 101, 2, '', '', '2020-01-08 21:03:13'),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29'),
 (8, 102, 1, '', '', '2020-01-09 23:54:33'),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (10, 104, 1, '', '', '2020-01-11 18:34:49'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49')]

### Runners Table

In [66]:
%%sql
CREATE TABLE runners (
  "runner_id" INTEGER,
  "registration_date" DATE
);

INSERT INTO runners
  ("runner_id", "registration_date")
VALUES
  (1, '2021-01-01'),
  (2, '2021-01-03'),
  (3, '2021-01-08'),
  (4, '2021-01-15');

 * sqlite://


'Connected: @None'

In [67]:
%%sql
SELECT *
FROM runners

 * sqlite://
Done.


[(1, '2021-01-01'), (2, '2021-01-03'), (3, '2021-01-08'), (4, '2021-01-15')]

### Runner Orders Table

In [68]:
%%sql
CREATE TABLE runner_orders (
  "order_id" INTEGER,
  "runner_id" INTEGER,
  "pickup_time" VARCHAR(19),
  "distance" VARCHAR(7),
  "duration" VARCHAR(10),
  "cancellation" VARCHAR(23)
);

INSERT INTO runner_orders
  ("order_id", "runner_id", "pickup_time", "distance", "duration", "cancellation")
VALUES
  ('1', '1', '2020-01-01 18:15:34', '20km', '32 minutes', ''),
  ('2', '1', '2020-01-01 19:10:54', '20km', '27 minutes', ''),
  ('3', '1', '2020-01-03 00:12:37', '13.4km', '20 mins', NULL),
  ('4', '2', '2020-01-04 13:53:03', '23.4', '40', NULL),
  ('5', '3', '2020-01-08 21:10:57', '10', '15', NULL),
  ('6', '3', 'null', 'null', 'null', 'Restaurant Cancellation'),
  ('7', '2', '2020-01-08 21:30:45', '25km', '25mins', 'null'),
  ('8', '2', '2020-01-10 00:15:02', '23.4 km', '15 minute', 'null'),
  ('9', '2', 'null', 'null', 'null', 'Customer Cancellation'),
  ('10', '1', '2020-01-11 18:50:20', '10km', '10minutes', 'null');

 * sqlite://
(sqlite3.OperationalError) table runner_orders already exists
[SQL: CREATE TABLE runner_orders (
  "order_id" INTEGER,
  "runner_id" INTEGER,
  "pickup_time" VARCHAR(19),
  "distance" VARCHAR(7),
  "duration" VARCHAR(10),
  "cancellation" VARCHAR(23)
);]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


In [69]:
%%sql
SELECT * 
FROM runner_orders

 * sqlite://
Done.


[(1, 1, '2020-01-01 18:15:34', '20km', '32 minutes', ''),
 (2, 1, '2020-01-01 19:10:54', '20km', '27 minutes', ''),
 (3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', ''),
 (4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (5, 3, '2020-01-08 21:10:57', '10', '15', ''),
 (6, 3, 'null', 'null', 'null', 'cancelled'),
 (7, 2, '2020-01-08 21:30:45', '25km', '25mins', ''),
 (8, 2, '2020-01-10 00:15:02', '23.4 km', '15 minute', ''),
 (9, 2, 'null', 'null', 'null', 'cancelled'),
 (10, 1, '2020-01-11 18:50:20', '10km', '10minutes', '')]

### Pizza Names Table

In [70]:
%%sql
CREATE TABLE pizza_names (
  "pizza_id" INTEGER,
  "pizza_name" TEXT
);

INSERT INTO pizza_names
  ("pizza_id", "pizza_name")
VALUES
  (1, 'Meatlovers'),
  (2, 'Vegetarian');

 * sqlite://
(sqlite3.OperationalError) table pizza_names already exists
[SQL: CREATE TABLE pizza_names (
  "pizza_id" INTEGER,
  "pizza_name" TEXT
);]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


In [71]:
%%sql
SELECT *
FROM pizza_names

 * sqlite://
Done.


[(1, 'Meatlovers'), (2, 'Vegetarian')]

### Pizza Recipes Table

In [72]:
%%sql
CREATE TABLE pizza_recipes (
  "pizza_id" INTEGER,
  "toppings" TEXT
);

INSERT INTO pizza_recipes
  ("pizza_id", "toppings")
VALUES
  (1, '1, 2, 3, 4, 5, 6, 8, 10'),
  (2, '4, 6, 7, 9, 11, 12');

 * sqlite://
(sqlite3.OperationalError) table pizza_recipes already exists
[SQL: CREATE TABLE pizza_recipes (
  "pizza_id" INTEGER,
  "toppings" TEXT
);]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


In [73]:
%%sql
SELECT *
FROM pizza_recipes

 * sqlite://
Done.


[(1, '1, 2, 3, 4, 5, 6, 8, 10'), (2, '4, 6, 7, 9, 11, 12')]

### Pizza Toppings Table

In [74]:
%%sql
DROP TABLE IF EXISTS pizza_toppings;
CREATE TABLE pizza_toppings (
  "topping_id" INTEGER,
  "topping_name" TEXT
);

INSERT INTO pizza_toppings
  ("topping_id", "topping_name")
VALUES
  (1, 'Bacon'),
  (2, 'BBQ Sauce'),
  (3, 'Beef'),
  (4, 'Cheese'),
  (5, 'Chicken'),
  (6, 'Mushrooms'),
  (7, 'Onions'),
  (8, 'Pepperoni'),
  (9, 'Peppers'),
  (10, 'Salami'),
  (11, 'Tomatoes'),
  (12, 'Tomato Sauce');

 * sqlite://
Done.
Done.
12 rows affected.


[]

In [75]:
%%sql
SELECT *
FROM pizza_toppings

 * sqlite://
Done.


[(1, 'Bacon'),
 (2, 'BBQ Sauce'),
 (3, 'Beef'),
 (4, 'Cheese'),
 (5, 'Chicken'),
 (6, 'Mushrooms'),
 (7, 'Onions'),
 (8, 'Pepperoni'),
 (9, 'Peppers'),
 (10, 'Salami'),
 (11, 'Tomatoes'),
 (12, 'Tomato Sauce')]

## Part 1: PIZZA METRICS

### 1. How many pizzas were ordered?

In [76]:
%%sql
SELECT
    COUNT(pizza_id) as pizzas_ordered
FROM customer_orders

 * sqlite://
Done.


[(14,)]

#### Observations

* 14 pizzas were ordered in total

### 2. How many unique customer orders were made?

In [77]:
%%sql
SELECT
    COUNT(DISTINCT order_id) as unique_customer_orders
FROM customer_orders 

 * sqlite://
Done.


[(10,)]

#### Observations

* 10 unique customer orders were made.

### 3. How many successful orders were delivered by each runner?

In [78]:
%%sql
SELECT * 
FROM runner_orders

 * sqlite://
Done.


[(1, 1, '2020-01-01 18:15:34', '20km', '32 minutes', ''),
 (2, 1, '2020-01-01 19:10:54', '20km', '27 minutes', ''),
 (3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', ''),
 (4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (5, 3, '2020-01-08 21:10:57', '10', '15', ''),
 (6, 3, 'null', 'null', 'null', 'cancelled'),
 (7, 2, '2020-01-08 21:30:45', '25km', '25mins', ''),
 (8, 2, '2020-01-10 00:15:02', '23.4 km', '15 minute', ''),
 (9, 2, 'null', 'null', 'null', 'cancelled'),
 (10, 1, '2020-01-11 18:50:20', '10km', '10minutes', '')]

#### Clean runner_orders table

- Replace 'None' and 'null' with blanks
- Replace any type of cancellation with the word 'cancelled' in the cancellation column

In [79]:
%%sql
UPDATE runner_orders
SET cancellation = 
    CASE 
        WHEN cancellation IS NULL OR cancellation = 'null' THEN ''
        WHEN cancellation LIKE '%Cancellation%' THEN 'cancelled'
        ELSE cancellation
    END;

 * sqlite://
10 rows affected.


[]

In [80]:
%%sql
SELECT *
FROM runner_orders
ORDER BY 2

 * sqlite://
Done.


[(1, 1, '2020-01-01 18:15:34', '20km', '32 minutes', ''),
 (2, 1, '2020-01-01 19:10:54', '20km', '27 minutes', ''),
 (3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', ''),
 (10, 1, '2020-01-11 18:50:20', '10km', '10minutes', ''),
 (4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (7, 2, '2020-01-08 21:30:45', '25km', '25mins', ''),
 (8, 2, '2020-01-10 00:15:02', '23.4 km', '15 minute', ''),
 (9, 2, 'null', 'null', 'null', 'cancelled'),
 (5, 3, '2020-01-08 21:10:57', '10', '15', ''),
 (6, 3, 'null', 'null', 'null', 'cancelled')]

In [81]:
%%sql
SELECT
    runner_id,
    COUNT(order_id) as successful_deliveries
FROM runner_orders
WHERE cancellation = ''
GROUP BY 1

 * sqlite://
Done.


[(1, 4), (2, 3), (3, 1)]

#### Observations
* Runner 1 had 4 successful deliveries
* Runner 2 had 3 successful deliveries
* Runner 3 had 1 successful delivery

### 4. How many of each type of pizza was delivered?

In [82]:
%%sql
SELECT
    pn.pizza_name,
    COUNT(*) as pizzas_delivered
FROM customer_orders as co
JOIN pizza_names as pn
ON co.pizza_id = pn.pizza_id
JOIN runner_orders as ro
ON co.order_id = ro.order_id
WHERE ro.cancellation = ''
GROUP BY 1

 * sqlite://
Done.


[('Meatlovers', 9), ('Vegetarian', 3)]

#### Observations

* 9 Meatlovers and 3 Vegetarian pizzas were delivered.

### 5. How many Vegetarian and Meatlovers were ordered by each customer?

In [83]:
%%sql
SELECT *
FROM customer_orders
ORDER BY 2

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02'),
 (2, 101, 1, '', '', '2020-01-01 19:00:52'),
 (6, 101, 2, '', '', '2020-01-08 21:03:13'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23'),
 (3, 102, 2, '', '', '2020-01-02 23:51:23'),
 (8, 102, 1, '', '', '2020-01-09 23:54:33'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46'),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29'),
 (10, 104, 1, '', '', '2020-01-11 18:34:49'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49'),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29')]

In [84]:
%%sql
SELECT 
    co.customer_id, 
    pn.pizza_name, 
    COUNT(co.order_id) AS pizzas_ordered
FROM customer_orders co
JOIN pizza_names as pn 
ON co.pizza_id = pn.pizza_id
GROUP BY 1, 2

 * sqlite://
Done.


[(101, 'Meatlovers', 2),
 (101, 'Vegetarian', 1),
 (102, 'Meatlovers', 2),
 (102, 'Vegetarian', 1),
 (103, 'Meatlovers', 3),
 (103, 'Vegetarian', 1),
 (104, 'Meatlovers', 3),
 (105, 'Vegetarian', 1)]

#### Observations
* Customer 101 ordered 2 Meatlovers and 1 Vegetarian
* Customer 102 ordered 2 Meatlovers and 1 Vegetarian
* Customer 103 ordered 3 Meatlovers and 1 Vegetarian
* Customer 104 ordered 3 Meatlovers
* Customer 105 ordered 1 Vegetarian

### 6. What was the maximum number of pizzas delivered in a single order?

#### Step 1: Count the number of pizzas in each order.

In [85]:
%%sql
SELECT
    order_id,
    COUNT(pizza_id) as cnt
FROM customer_orders as co
GROUP BY 1

 * sqlite://
Done.


[(1, 1),
 (2, 1),
 (3, 2),
 (4, 3),
 (5, 1),
 (6, 1),
 (7, 1),
 (8, 1),
 (9, 1),
 (10, 2)]

#### Step 2: Determine which order had the most pizzas delivered

In [86]:
%%sql
WITH pizzas_per_order AS (
    SELECT
        co.order_id,
        COUNT(*) as pizza_count
    FROM customer_orders as co
    GROUP BY 1
)

SELECT
    ppo.order_id,
    MAX(ppo.pizza_count) as pizzas_deliverd
FROM pizzas_per_order as ppo
JOIN runner_orders as ro
ON ppo.order_id = ro.order_id
WHERE ro.cancellation = ''

 * sqlite://
Done.


[(4, 3)]

#### Observations
* Order number 4 had the most pizzas delivered in a single order with 3.

### 7. For each customer, how many delivered pizzas had at least 1 change and how many had no changes?

In [87]:
%%sql
SELECT *
FROM customer_orders

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02'),
 (2, 101, 1, '', '', '2020-01-01 19:00:52'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23'),
 (3, 102, 2, '', '', '2020-01-02 23:51:23'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46'),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29'),
 (6, 101, 2, '', '', '2020-01-08 21:03:13'),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29'),
 (8, 102, 1, '', '', '2020-01-09 23:54:33'),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (10, 104, 1, '', '', '2020-01-11 18:34:49'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49')]

#### Clean customer orders table
* Replace Null values with blanks in Exclusions column

In [88]:
%%sql
UPDATE customer_orders
SET exclusions = 
    CASE 
        WHEN exclusions = 'null' THEN ''
        ELSE exclusions
    END;

 * sqlite://
14 rows affected.


[]

* Replace Null values with blanks in Extras column

In [89]:
%%sql
UPDATE customer_orders
SET extras = 
    CASE
      WHEN extras IS NULL or extras = 'null' THEN ''
      ELSE extras
END;

 * sqlite://
14 rows affected.


[]

In [90]:
%%sql
SELECT *
FROM customer_orders as co
JOIN runner_orders as ro
ON co.order_id = ro.order_id
ORDER BY 2

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02', 1, 1, '2020-01-01 18:15:34', '20km', '32 minutes', ''),
 (2, 101, 1, '', '', '2020-01-01 19:00:52', 2, 1, '2020-01-01 19:10:54', '20km', '27 minutes', ''),
 (6, 101, 2, '', '', '2020-01-08 21:03:13', 6, 3, 'null', 'null', 'null', 'cancelled'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23', 3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', ''),
 (3, 102, 2, '', '', '2020-01-02 23:51:23', 3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', ''),
 (8, 102, 1, '', '', '2020-01-09 23:54:33', 8, 2, '2020-01-10 00:15:02', '23.4 km', '15 minute', ''),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59', 9, 2, 'null', 'null', 'null', 'cancelled'),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29', 5, 3, '2020-01-08 21:10:57', '10', '15', ''),
 (10, 104, 1, '', '', '2020-01-11 18:34:49', 10, 1, '2020-01-11 18:50:20', '10km', '10minutes', ''),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49', 10, 1, '2020-01-11 18:50:20', '10km', '10minutes', ''),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29', 7, 2, '2020-01-08 21:30:45', '25km', '25mins', '')]

In [91]:
%%sql
SELECT 
    co.customer_id,
    COUNT(CASE WHEN co.exclusions = '' AND co.extras = '' THEN 1 END) as no_change,
    COUNT(CASE WHEN co.exclusions != '' OR co.extras != '' THEN 1 END) as has_change
FROM customer_orders as co
JOIN runner_orders as ro
ON co.order_id = ro.order_id
WHERE ro.cancellation = ''
GROUP BY 1

 * sqlite://
Done.


[(101, 2, 0), (102, 3, 0), (103, 0, 3), (104, 1, 2), (105, 0, 1)]

#### Observations
* Customer 101 had 2 pizzas delivered without changes and 0 with changes
* Customer 102 had 3 pizzas delivered without changes and 0 with changes
* Customer 103 had 3 pizzas delivered with changes and 0 without any changes
* Customer 104 had 1 pizza delivered without any changes and 2 with changes
* Customer 105 had 1 pizza deliverd with a change and 0 without any

### 8. How many pizzas were delivered that had both exclusions and extras?

In [92]:
%%sql
SELECT *
FROM customer_orders
WHERE exclusions != '' AND extras != ''

 * sqlite://
Done.


[(9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49')]

In [93]:
%%sql
SELECT
    COUNT(*) as exclusions_and_extras
FROM customer_orders as co
JOIN runner_orders as ro
ON co.order_id = ro.order_id
WHERE exclusions != '' AND extras != '' AND ro.cancellation = ''

 * sqlite://
Done.


[(1,)]

#### Observations
* Only 1 pizza was delivered that had both exclusions and extras
* Another pizza with exclusions and extras was ordered but was cancelled before it was delivered

### 9. What was the total volume of pizzas ordered for each hour of the day?

In [94]:
%%sql
SELECT 
    strftime('%H', order_time) as hour,
    COUNT(*) as pizzas_ordered
FROM customer_orders
GROUP BY 1

 * sqlite://
Done.


[('11', 1), ('13', 3), ('18', 3), ('19', 1), ('21', 3), ('23', 3)]

#### Observations
* 1 pizza was ordered between 11 am and 12 pm
* 3 pizzas were ordered between 1 and 2 pm 
* 3 pizzas were ordered between 6 and 7 pm
* 1 pizza was ordered between 7 and 8 pm
* 3 pizzas were ordered between 9 and 10 pm
* 2 pizzas were ordered between 11 and 12 pm

### 10. What was the volume of orders for each day of the week?

In [95]:
%%sql
SELECT *
FROM customer_orders

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02'),
 (2, 101, 1, '', '', '2020-01-01 19:00:52'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23'),
 (3, 102, 2, '', '', '2020-01-02 23:51:23'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46'),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29'),
 (6, 101, 2, '', '', '2020-01-08 21:03:13'),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29'),
 (8, 102, 1, '', '', '2020-01-09 23:54:33'),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (10, 104, 1, '', '', '2020-01-11 18:34:49'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49')]

In [96]:
%%sql
SELECT 
    strftime('%w', order_time) as day_of_week,
    COUNT(DISTINCT order_id) as number_of_orders
FROM customer_orders
GROUP BY 1

 * sqlite://
Done.


[('3', 5), ('4', 2), ('5', 1), ('6', 2)]

##### Notes
* Day of Week = 3 (Wednesday)

#### Observations
* 5 orders on Wednesday
* 2 on Thursday 
* 1 on Friday
* 2 on Saturday

## Part 2: Runner and Customer Experience

### 1. How many runners signed up for each one week period? (i.e. week starts 2021-01-01)

In [97]:
%%sql
SELECT *
FROM runners

 * sqlite://
Done.


[(1, '2021-01-01'), (2, '2021-01-03'), (3, '2021-01-08'), (4, '2021-01-15')]

In [98]:
%%sql
SELECT 
  strftime('%Y-%m-%d', registration_date, 'weekday 4', '-6 days') AS week_start,
  COUNT(*) AS total_runners
FROM runners
GROUP BY 1
ORDER BY 2 DESC, 1

 * sqlite://
Done.


[('2021-01-01', 2), ('2021-01-08', 1), ('2021-01-15', 1)]

#### Observations
* 2 runners started week of 2021-01-01
* 1 runner started week of 2021-01-08
* 1 runner started week of 2021-01-15

### 2. What was the average time in minutes it took for each runner to pickup the order?

In [99]:
%%sql
SELECT
        *,
        ROUND((strftime('%s', ro.pickup_time) - strftime('%s', co.order_time)) / 60.0, 2) AS time_to_pickup
    FROM customer_orders as co
    JOIN runner_orders as ro
    ON co.order_id = ro.order_id
    WHERE ro.cancellation = ''
    ORDER BY ro.runner_id

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02', 1, 1, '2020-01-01 18:15:34', '20km', '32 minutes', '', 10.53),
 (2, 101, 1, '', '', '2020-01-01 19:00:52', 2, 1, '2020-01-01 19:10:54', '20km', '27 minutes', '', 10.03),
 (3, 102, 1, '', '', '2020-01-02 23:51:23', 3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', '', 21.23),
 (3, 102, 2, '', '', '2020-01-02 23:51:23', 3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', '', 21.23),
 (10, 104, 1, '', '', '2020-01-11 18:34:49', 10, 1, '2020-01-11 18:50:20', '10km', '10minutes', '', 15.52),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49', 10, 1, '2020-01-11 18:50:20', '10km', '10minutes', '', 15.52),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', '', 29.28),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', '', 29.28),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', '', 29.28),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29', 7, 2, '2020-01-08 21:30:45', '25km', '25mins', '', 10.27),
 (8, 102, 1, '', '', '2020-01-09 23:54:33', 8, 2, '2020-01-10 00:15:02', '23.4 km', '15 minute', '', 20.48),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29', 5, 3, '2020-01-08 21:10:57', '10', '15', '', 10.47)]

In [100]:
%%sql
WITH time_to_pickup AS (
    SELECT
        *,
        ROUND((strftime('%s', ro.pickup_time) - strftime('%s', co.order_time)) / 60.0, 2) AS time_to_pickup
    FROM customer_orders as co
    JOIN runner_orders as ro
    ON co.order_id = ro.order_id
    WHERE ro.cancellation = ''
    ORDER BY ro.runner_id
)

SELECT 
    runner_id,
    ROUND(AVG(DISTINCT time_to_pickup),2) as time_to_pickup
FROM time_to_pickup
GROUP BY 1

 * sqlite://
Done.


[(1, 14.33), (2, 20.01), (3, 10.47)]

#### Observations
* Runner 1 took on average 14.33 minutes to pickup an order
* Runner 2 took on average 20.01 minutes to pickup an order
* Runner 3 took on average 10.47 minutes (only delivered 1 order)

### 3. Is there any relationship between the number of pizzas and how long the order takes to prepare?

In [101]:
%%sql
SELECT
  co.order_id,
  COUNT(*) as pizzas_ordered,
  ROUND((strftime('%s', ro.pickup_time) - strftime('%s', co.order_time)) / 60.0, 2) AS preparation_time
FROM customer_orders as co
JOIN runner_orders as ro 
ON co.order_id = ro.order_id
WHERE ro.cancellation = ''
GROUP BY 1

 * sqlite://
Done.


[(1, 1, 10.53),
 (2, 1, 10.03),
 (3, 2, 21.23),
 (4, 3, 29.28),
 (5, 1, 10.47),
 (7, 1, 10.27),
 (8, 1, 20.48),
 (10, 2, 15.52)]

#### Observations
* The first 6 orders seem to indicate that each pizza takes about 10-11 minutes to prepare.
* The last 2 orders tell a different story, however. Order_id 8 only had 1 pizza ordered but took over 20 minutes to prepare, while order_id 10 had 2 pizzas ordered and only took 15.5 minutes to prepare.

### 4. What was the average distance travelled for each customer?

#### Clean Runner Orders Table
* Replace 'km' in distance column with blanks

In [102]:
%%sql
UPDATE runner_orders
SET distance = REPLACE(distance, 'km', '')

 * sqlite://
10 rows affected.


[]

In [103]:
%%sql
SELECT *
FROM customer_orders as co
JOIN runner_orders as ro
ON co.order_id = ro.order_id
WHERE ro.cancellation = ''
ORDER BY customer_id

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02', 1, 1, '2020-01-01 18:15:34', '20', '32 minutes', ''),
 (2, 101, 1, '', '', '2020-01-01 19:00:52', 2, 1, '2020-01-01 19:10:54', '20', '27 minutes', ''),
 (3, 102, 1, '', '', '2020-01-02 23:51:23', 3, 1, '2020-01-03 00:12:37', '13.4', '20 mins', ''),
 (3, 102, 2, '', '', '2020-01-02 23:51:23', 3, 1, '2020-01-03 00:12:37', '13.4', '20 mins', ''),
 (8, 102, 1, '', '', '2020-01-09 23:54:33', 8, 2, '2020-01-10 00:15:02', '23.4 ', '15 minute', ''),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46', 4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29', 5, 3, '2020-01-08 21:10:57', '10', '15', ''),
 (10, 104, 1, '', '', '2020-01-11 18:34:49', 10, 1, '2020-01-11 18:50:20', '10', '10minutes', ''),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49', 10, 1, '2020-01-11 18:50:20', '10', '10minutes', ''),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29', 7, 2, '2020-01-08 21:30:45', '25', '25mins', '')]

In [104]:
%%sql
SELECT
    co.customer_id,
    ROUND(AVG(DISTINCT ro.distance),2) as avg_distance
FROM customer_orders as co
JOIN runner_orders as ro
ON co.order_id = ro.order_id
WHERE ro.cancellation = ''
GROUP BY 1

 * sqlite://
Done.


[(101, 20.0), (102, 18.4), (103, 23.4), (104, 10.0), (105, 25.0)]

#### Observations
* Average distance traveled for customer 101 was 20 km
* Average distance travelled for customer 102 was 18.4 km 
* Average distance travelled for customer 103 was 23.4 km
* Average distance travelled for customer 104 was 10 km
* Average distance travelled for customer 105 was 25 km

### 5. What was the difference between the longest and shortest delivery times for all orders?

#### Clean Runner Orders Table
* Remove minutes from duration column

In [105]:
%%sql
UPDATE runner_orders
SET duration = REPLACE(REPLACE(REPLACE(duration, 'minutes', ''), 'mins', ''), 'minute', '');

 * sqlite://
10 rows affected.


[]

In [106]:
%%sql
SELECT *
FROM runner_orders

 * sqlite://
Done.


[(1, 1, '2020-01-01 18:15:34', '20', '32 ', ''),
 (2, 1, '2020-01-01 19:10:54', '20', '27 ', ''),
 (3, 1, '2020-01-03 00:12:37', '13.4', '20 ', ''),
 (4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (5, 3, '2020-01-08 21:10:57', '10', '15', ''),
 (6, 3, 'null', 'null', 'null', 'cancelled'),
 (7, 2, '2020-01-08 21:30:45', '25', '25', ''),
 (8, 2, '2020-01-10 00:15:02', '23.4 ', '15 ', ''),
 (9, 2, 'null', 'null', 'null', 'cancelled'),
 (10, 1, '2020-01-11 18:50:20', '10', '10', '')]

In [107]:
%%sql
SELECT
    MAX(duration) AS slowest_delivery,
    MIN(duration) AS fastest_delivery,
    MAX(duration) - MIN(duration) AS fastest_slowest_diff
FROM runner_orders
WHERE cancellation = ''

 * sqlite://
Done.


[('40', '10', 30)]

#### Observations
* The slowest delivery took 40 minutes
* The fastest delivery took 10 minutes
* The difference between the fastest and slowest deliveries was 30 minutes

### 6. What was the average speed for each runner?

In [108]:
%%sql
WITH runner_speed AS (
    SELECT
        runner_id,
        distance,
        duration,
        ROUND((1.0 * distance) / duration, 2) as km_min
    FROM runner_orders
    WHERE cancellation = ''
    ORDER BY 1
)

SELECT
    runner_id,
    ROUND(AVG(km_min),2) as average_speed
FROM runner_speed
GROUP BY 1

 * sqlite://
Done.


[(1, 0.76), (2, 1.05), (3, 0.67)]

#### Observations

* The average speed of runner 1 was 0.76 km/min
* The average speed of runner 2 was 1.05 km/min
* The average speed of runner 3 was 0.67 km min

## Part 3: Ingredient Optimization

### 1. What are the standard ingredients for each pizza?

In [131]:
%%sql
SELECT *
FROM pizza_names

 * sqlite://
Done.


[(1, 'Meatlovers'), (2, 'Vegetarian')]

In [133]:
%%sql
SELECT *
FROM pizza_recipes

 * sqlite://
Done.


[(1, '1, 2, 3, 4, 5, 6, 8, 10'), (2, '4, 6, 7, 9, 11, 12')]

In [134]:
%%sql
SELECT 
    pn.pizza_name,
    pr.toppings
FROM pizza_names as pn
JOIN pizza_recipes as pr
ON pn.pizza_id = pr.pizza_id

 * sqlite://
Done.


[('Meatlovers', '1, 2, 3, 4, 5, 6, 8, 10'),
 ('Vegetarian', '4, 6, 7, 9, 11, 12')]

In [132]:
%%sql
SELECT *
FROM pizza_toppings

 * sqlite://
Done.


[(1, 'Bacon'),
 (2, 'BBQ Sauce'),
 (3, 'Beef'),
 (4, 'Cheese'),
 (5, 'Chicken'),
 (6, 'Mushrooms'),
 (7, 'Onions'),
 (8, 'Pepperoni'),
 (9, 'Peppers'),
 (10, 'Salami'),
 (11, 'Tomatoes'),
 (12, 'Tomato Sauce')]

#### Observations
* The standard toppings for Meatlovers are Bacon, BBQ Sauce, Beef, Cheese, Chicken, Mushrooms, Pepperoni and Salami
* The standard toppings for Vegetarian are Cheese, Mushrooms, Onions, Peppers, Tomatoes and Tomato Sauce

### 2. What was the most commonly added extra?

In [142]:
%%sql
SELECT *
FROM customer_orders as co

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02'),
 (2, 101, 1, '', '', '2020-01-01 19:00:52'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23'),
 (3, 102, 2, '', '', '2020-01-02 23:51:23'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46'),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29'),
 (6, 101, 2, '', '', '2020-01-08 21:03:13'),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29'),
 (8, 102, 1, '', '', '2020-01-09 23:54:33'),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (10, 104, 1, '', '', '2020-01-11 18:34:49'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49')]

### Observations
* The most commonly added extra is bacon

### 3. What was the most common exclusion?

In [145]:
%%sql
SELECT *
FROM customer_orders

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02'),
 (2, 101, 1, '', '', '2020-01-01 19:00:52'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23'),
 (3, 102, 2, '', '', '2020-01-02 23:51:23'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46'),
 (5, 104, 1, '', '1', '2020-01-08 21:00:29'),
 (6, 101, 2, '', '', '2020-01-08 21:03:13'),
 (7, 105, 2, '', '1', '2020-01-08 21:20:29'),
 (8, 102, 1, '', '', '2020-01-09 23:54:33'),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (10, 104, 1, '', '', '2020-01-11 18:34:49'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49')]

#### Observations
* The most common exclusion was cheese

## Part 4: Pricing and Ratings

### 1. Meat Lovers pizza costs 12 dollars and Vegetarian costs 10 dollars. If there are no delivery fees and no charges for changes, how much money has Pizza Runner made so far?

In [48]:
%%sql
SELECT
    SUM(CASE WHEN pn.pizza_name = 'Meatlovers' THEN 12 END) as revenue_meat,
    SUM(CASE WHEN pn.pizza_name = 'Vegetarian' THEN 10 END) as revenue_veg,
    SUM(
      CASE
        WHEN pn.pizza_name = 'Meatlovers' THEN 12
        WHEN pn.pizza_name = 'Vegetarian' THEN 10
        ELSE 0
      END
    ) as total_revenue
FROM customer_orders as co
JOIN pizza_names as pn
ON co.pizza_id = pn.pizza_id
JOIN runner_orders as ro
ON ro.order_id = co.order_id
WHERE ro.cancellation = ''

 * sqlite://
Done.


[(108, 30, 138)]

#### Observations
* Pizza Diner has made 138 dollars so far. 
* 108 dollars from Meatlovers and 30 dollars from Vegetarian.

### 2. What if there was an additional $1 charge for any extras that were added?

In [49]:
%%sql
SELECT
    SUM (
      CASE
        WHEN pn.pizza_name = 'Meatlovers' THEN 12
        WHEN pn.pizza_name = 'Vegetarian' THEN 10
        ELSE 0
    END +
      CASE
        WHEN co.extras = '1' THEN 1
        WHEN co.extras = '1, 4' THEN 2
        ELSE 0
    END
    ) as total_revenue
FROM customer_orders as co
JOIN pizza_names as pn
ON co.pizza_id = pn.pizza_id
JOIN runner_orders as ro
ON ro.order_id = co.order_id
WHERE ro.cancellation = ''

 * sqlite://
Done.


[(142,)]

#### Observations
* Pizza Runner would have made 142 dollars so far if each extra that was added cost 1 dollar
* In total, there were 4 extras added to pizzas which were were successfully delivered

### 3. The Pizza Runner team now wants to add an additional ratings system that allows customers to rate their runner, how would you design an additional table for this new dataset - generate a schema for this new table and insert your own data for ratings for each successful customer order between 1 to 5.

In [50]:
%%sql
SELECT *
FROM runner_orders

 * sqlite://
Done.


[(1, 1, '2020-01-01 18:15:34', '20', '32 ', ''),
 (2, 1, '2020-01-01 19:10:54', '20', '27 ', ''),
 (3, 1, '2020-01-03 00:12:37', '13.4', '20 ', ''),
 (4, 2, '2020-01-04 13:53:03', '23.4', '40', ''),
 (5, 3, '2020-01-08 21:10:57', '10', '15', ''),
 (6, 3, 'null', 'null', 'null', 'cancelled'),
 (7, 2, '2020-01-08 21:30:45', '25', '25', ''),
 (8, 2, '2020-01-10 00:15:02', '23.4 ', '15 ', ''),
 (9, 2, 'null', 'null', 'null', 'cancelled'),
 (10, 1, '2020-01-11 18:50:20', '10', '10', '')]

In [51]:
%%sql
CREATE TABLE order_ratings (
  "order_id" INTEGER,
  "rating" INTEGER
);

 * sqlite://
Done.


[]

In [52]:
%%sql
INSERT INTO order_ratings ("order_id", "rating")
VALUES
  (1, 3),
  (2, 4),
  (3, 4),
  (4, 2),
  (5, 4),
  (6, null),
  (7, 4),
  (8, 5),
  (9, null),
  (10, 5);

 * sqlite://
10 rows affected.


[]

In [53]:
%%sql
SELECT *
FROM order_ratings

 * sqlite://
Done.


[(1, 3),
 (2, 4),
 (3, 4),
 (4, 2),
 (5, 4),
 (6, None),
 (7, 4),
 (8, 5),
 (9, None),
 (10, 5)]

In [54]:
%%sql
SELECT *
FROM runner_orders as ro
JOIN order_ratings
ON ro.order_id = order_ratings.order_id
WHERE cancellation = ''
ORDER BY runner_id

 * sqlite://
Done.


[(1, 1, '2020-01-01 18:15:34', '20', '32 ', '', 1, 3),
 (2, 1, '2020-01-01 19:10:54', '20', '27 ', '', 2, 4),
 (3, 1, '2020-01-03 00:12:37', '13.4', '20 ', '', 3, 4),
 (10, 1, '2020-01-11 18:50:20', '10', '10', '', 10, 5),
 (4, 2, '2020-01-04 13:53:03', '23.4', '40', '', 4, 2),
 (7, 2, '2020-01-08 21:30:45', '25', '25', '', 7, 4),
 (8, 2, '2020-01-10 00:15:02', '23.4 ', '15 ', '', 8, 5),
 (5, 3, '2020-01-08 21:10:57', '10', '15', '', 5, 4)]

In [55]:
%%sql
SELECT 
    runner_id,
    ROUND(AVG(rating), 2) as avg_rating
FROM runner_orders as ro
JOIN order_ratings
ON ro.order_id = order_ratings.order_id
WHERE cancellation = ''
GROUP BY 1

 * sqlite://
Done.


[(1, 4.0), (2, 3.67), (3, 4.0)]

#### Observations
* Runner 1 and 3 both had an average rating of 4.0
* Runner 2 had an average rating of 3.67

### 4. Form a table which has the following information for successful deliveries.

* customer_id
* order_id
* runner_id
* rating
* order_time
* pickup_time

In [56]:
%%sql
SELECT
    co.customer_id,
    co.order_id,
    ro.runner_id,
    order_ratings.rating,
    co.order_time,
    ro.pickup_time    
FROM customer_orders as co
JOIN runner_orders as ro
ON co.order_id = ro.order_id
JOIN order_ratings
ON co.order_id = order_ratings.order_id
WHERE ro.cancellation = ''
ORDER BY ro.runner_id

 * sqlite://
Done.


[(101, 1, 1, 3, '2020-01-01 18:05:02', '2020-01-01 18:15:34'),
 (101, 2, 1, 4, '2020-01-01 19:00:52', '2020-01-01 19:10:54'),
 (102, 3, 1, 4, '2020-01-02 23:51:23', '2020-01-03 00:12:37'),
 (102, 3, 1, 4, '2020-01-02 23:51:23', '2020-01-03 00:12:37'),
 (104, 10, 1, 5, '2020-01-11 18:34:49', '2020-01-11 18:50:20'),
 (104, 10, 1, 5, '2020-01-11 18:34:49', '2020-01-11 18:50:20'),
 (103, 4, 2, 2, '2020-01-04 13:23:46', '2020-01-04 13:53:03'),
 (103, 4, 2, 2, '2020-01-04 13:23:46', '2020-01-04 13:53:03'),
 (103, 4, 2, 2, '2020-01-04 13:23:46', '2020-01-04 13:53:03'),
 (105, 7, 2, 4, '2020-01-08 21:20:29', '2020-01-08 21:30:45'),
 (102, 8, 2, 5, '2020-01-09 23:54:33', '2020-01-10 00:15:02'),
 (104, 5, 3, 4, '2020-01-08 21:00:29', '2020-01-08 21:10:57')]

### 5. If a Meat Lovers pizza costs 12 dollars and Vegetarian costs 10 dollars and there is no cost for extras and each runner is paid $0.30 per kilometre traveled - how much money does Pizza Runner have left over after these deliveries?

In [61]:
%%sql
WITH total_revenue AS (
SELECT
    SUM(
      CASE
        WHEN pn.pizza_name = 'Meatlovers' THEN 12
        WHEN pn.pizza_name = 'Vegetarian' THEN 10
        ELSE 0
      END
    ) as total_revenue
FROM customer_orders as co
JOIN pizza_names as pn
ON co.pizza_id = pn.pizza_id
JOIN runner_orders as ro
ON ro.order_id = co.order_id
WHERE ro.cancellation = ''
),

amount_paid AS (
SELECT
    SUM(distance * 0.3) as amount_paid
FROM runner_orders
WHERE cancellation = ''
)

SELECT
    (tr.total_revenue - ap.amount_paid) as revenue_after_deliveries
FROM total_revenue as tr
JOIN amount_paid as ap


 * sqlite://
Done.


[(94.44,)]

#### Observations
* Pizza Runner has 94.44 dollars after the deliveries were made.

### Conclusions

* 14 total pizzas were ordered across 10 unique orders
* 2 orders were cancelled (order 6 and order 9)
* 12 total pizzas were successfully delivered across 8 unique orders
* Across the 8 orders which were successfully delivered, 3 had multiple pizzas
* Of the 12 pizzas that were deliverd, 9 were Meatlovers and only 3 were Vegetarian
* Meatlovers outsold Vegetarian 3 to 1 even though Meatlovers was 2 dollars more expensive
* The most common extra was bacon, while the most common exclusion was cheese
* The fastest delivery took only 10 minutes while the slowest delivery took 40 minutes
* Pizza Runner has made 138 dollars in revenue so far